File: R/events.R
##' Discrete events applied during an SCM run.
##'
##' An event is a \code{(time, action)} pair applied between solver legs: the
##' integrator is stopped at the event time, the action changes the patch, the
##' system is recomputed, and integration resumes from the new state. Node
##' introduction has always worked this way; these functions make the same
##' mechanism available for resource pulses, harvest and climate extremes.
##'
##' The vocabulary here is deliberately model-agnostic, because the machinery is
##' shared by every strategy and environment and nothing about it is specific to
##' plants: a resource pulse is water in TF24 and could be anything countable in
##' a size-structured animal model, and a climate extreme is heat in one model
##' and could be cold or salinity in another. Names that are only true of one
##' model live with that model — \code{\link{rainfall_pulse}} is a resource
##' pulse of water into TF24's surface soil layer.
##'
##' Each event carries when it happens, what kind of thing it is, what it acts
##' on, and the values it needs. The target is \code{"environment"} (its
##' resource pools and drivers), \code{"patch"} (every species) or
##' \code{"species"} (one, named by \code{species}). There is deliberately no
##' per-cohort target: a cohort has no stable address across a run, because
##' nodes are appended and never removed and schedule refinement changes how
##' many exist. Selecting particular cohorts is expressed as a size range
##' instead, which is both well defined and what size-selective removal needs.
##'
##' Events are instantaneous \emph{to the solver}: patch time does not advance
##' across one. An action is still free to reach its answer by integrating its
##' own fast sub-model over a nominal duration — \code{\link{climate_extreme}}
##' does — with demography frozen. The solver sees one jump either way.
##'
##' Events sharing a time are applied in a fixed order: environment events
##' first, then demographic ones, then node introductions — so a node
##' introduced at that instant starts life in the post-event environment.
##'
##' Because an event is also a stop time for the integrator, adding one changes
##' the adaptive step sequence. A run with events legitimately differs from one
##' without, at solver tolerance, even away from the events themselves.
##'
##' What each event actually did — as against what was asked of it — is
##' recorded, and readable afterwards as \code{scm$event_log}. The two differ
##' routinely: a pulse is capped at what the pool can hold, and harvesting a
##' size class removes whatever was in it.
##'
##' @param ... For \code{events}, objects returned by the individual event
##'   constructors, or whole \code{Events} objects; these are concatenated. Each
##'   constructor is vectorised over its arguments, so a whole series of
##'   pulses is one call.
##' @return An \code{Events} object: a list with \code{time}, \code{type},
##'   \code{target}, \code{target_index} and \code{params}, in schedule order.
##' @rdname events
##' @export
##' @examples
##' p <- scm_base_parameters("FF16")
##' p <- add_strategies(p, trait_matrix(1, "lma"))
##' ev <- events(
##'   events_default(p),
##'   harvest(time = 20, fraction = 0.3)
##' )
events <- function(...) {
  parts <- list(...)
  parts <- parts[!vapply(parts, is.null, logical(1))]
  if (length(parts) == 0L) {
    return(empty_events())
  }
  ev <- do.call(join_event_rows, parts)
  ## Sort the way the queue applies them -- by time, then by type -- so the
  ## object you read is in the order things actually happen. Sorting by time
  ## alone would list a different order from the one that runs, which is the
  ## kind of discrepancy nobody checks until it matters. `order` is stable, so
  ## events of one type at one instant keep the order they were given in, and
  ## that matters: two pulses at one time are capped in sequence against the
  ## same pool, so the order decides which is credited with the water.
  i <- order(ev$time, match(ev$type, event_type_order()))
  Events(time = as.numeric(ev$time[i]),
         type = as.character(ev$type[i]),
         target = as.character(ev$target[i]),
         target_index = as.integer(ev$target_index[i]),
         params = ev$params[i])
}

## The within-time application order, mirroring EventType in node_schedule.h:
## environment first (it sets the conditions), then removals, then node
## introduction last so a newborn starts in the post-event environment.
## Kept in step with the C++ enum by the round-trip test in test-events.R.
event_type_order <- function() {
  c("resource_pulse", "climate_extreme", "harvest", "node_introduction")
}

## The "no events supplied" signal: the C++ side falls back to
## p$node_schedule_times when the list is empty.
empty_events <- function() {
  Events(time = numeric(0), type = character(0), target = character(0),
         target_index = integer(0), params = list())
}

## Concatenate event rows. unlist() on an empty list gives NULL, which would
## silently turn an empty schedule into a malformed one, so the empty case is
## spelled out with the right types.
join_event_rows <- function(...) {
  parts <- list(...)
  pull <- function(field, empty) {
    if (length(parts) == 0L) return(empty)
    out <- unlist(lapply(parts, `[[`, field), use.names = FALSE)
    if (is.null(out)) empty else out
  }
  list(time = pull("time", numeric(0)),
       type = pull("type", character(0)),
       target = pull("target", character(0)),
       target_index = pull("target_index", integer(0)),
       params = if (length(parts) == 0L) list() else
         unlist(lapply(parts, `[[`, "params"), recursive = FALSE))
}

## Build one event type's rows. Vectorised over time and over each parameter;
## all must be length 1 or a common length n.
event_rows <- function(type, time, target, target_index = 1L, params = list()) {
  n <- max(c(length(time), vapply(params, length, integer(1)), 1L))
  recycle <- function(x, what) {
    if (length(x) == n) return(x)
    if (length(x) == 1L) return(rep(x, n))
    stop(sprintf("'%s' must be length 1 or %d, not %d", what, n, length(x)),
         call. = FALSE)
  }
  time <- recycle(time, "time")
  target <- recycle(target, "target")
  target_index <- recycle(target_index, "species")
  params <- lapply(seq_along(params),
                   function(j) recycle(params[[j]], names(params)[[j]]))
  list(time = as.numeric(time),
       type = rep(type, n),
       target = as.character(target),
       target_index = as.integer(target_index),
       ## One numeric vector per event, in the order the C++ action reads them.
       params = lapply(seq_len(n), function(i) {
         as.numeric(vapply(params, function(p) p[[i]], numeric(1)))
       }))
}

## "patch" unless a species was named, in which case "species".
scope_of <- function(species) {
  if (is.null(species)) "patch" else "species"
}

##' @details \code{events_default(p)} is the schedule a run gets when no events
##'   are supplied: the node introductions from \code{p$node_schedule_times} and
##'   nothing else. Start from it when adding events to an otherwise ordinary
##'   run — \code{events(events_default(p), harvest(...))} — or pass it
##'   on its own, which reproduces the default run exactly.
##' @rdname events
##' @export
events_default <- function(p) {
  events(node_introductions(p))
}

##' @param p Parameters object, whose \code{node_schedule_times} supply the
##'   introduction times.
##' @rdname events
##' @export
node_introductions <- function(p) {
  times <- p$node_schedule_times
  parts <- lapply(seq_along(times), function(i) {
    event_rows("node_introduction", time = times[[i]],
               target = "species", target_index = i)
  })
  do.call(join_event_rows, parts)
}

##' @param time Event time(s), in years of patch age.
##' @param resource Index of the environment's resource pool to add to. What a
##'   resource is depends on the environment: TF24's are its soil layers, so
##'   \code{resource = 1} is the surface. See \code{\link{rainfall_pulse}} for
##'   that case under a name that reads correctly.
##' @param amount How much to add, in whatever the environment measures that
##'   resource in. What the pool cannot hold is shed rather than forced in;
##'   \code{scm$event_log} reports how much of each.
##' @rdname events
##' @export
resource_pulse <- function(time, resource, amount) {
  event_rows("resource_pulse", time = time, target = "environment",
             target_index = resource, params = list(amount = amount))
}

##' @param fraction Fraction of individuals removed, in \code{[0, 1)}.
##' @param size_min,size_max Only individuals whose size falls in this band are
##'   removed. Size is the individual's size coordinate — height, for plant's
##'   models. The defaults take the whole population.
##' @param species Index of the species to act on; \code{NULL} (the default)
##'   acts on every species in the patch.
##' @details \code{harvest} covers removal generally, not only the forestry
##'   sense: leave the size band at its defaults and it is an across-the-board
##'   knock-down; set \code{size_min} and it takes everything above a size; set
##'   both and it thins one size class. They are one action, differing only in
##'   which individuals are selected.
##' @rdname events
##' @export
harvest <- function(time, fraction, size_min = 0, size_max = Inf,
                    species = NULL) {
  event_rows("harvest", time = time, target = scope_of(species),
             target_index = if (is.null(species)) 1L else species,
             params = list(fraction = fraction, size_min = size_min,
                           size_max = size_max))
}

##' @param intensity Peak level reached during the episode, in whatever the
##'   model measures the stressor in (deg C for a heatwave, say).
##' @param duration Nominal duration of the episode (years; a fortnight is
##'   \code{14 / 365}). The action sub-integrates over it at half-hourly steps;
##'   patch time does not advance.
##' @param threshold Level above which dose accrues; below it the episode does
##'   nothing.
##' @param sensitivity Dose accrued per unit-year above \code{threshold}.
##' @rdname events
##' @export
climate_extreme <- function(time, intensity, duration = 14 / 365,
                            threshold = 40, sensitivity = 1,
                            species = NULL) {
  event_rows("climate_extreme", time = time, target = scope_of(species),
             target_index = if (is.null(species)) 1L else species,
             params = list(intensity = intensity, duration = duration,
                           threshold = threshold,
                           sensitivity = sensitivity))
}

File: src/events.cpp
#include <plant/events.h>
#include <Rcpp.h>

namespace plant {

namespace {
// One table, read every way, so a name, its tag, its arity and what it may be
// aimed at cannot drift apart. Order follows EventType's, which is also the
// within-time application order.
struct TypeInfo {
  EventType type;
  const char* name;
  size_t n_params;
  EventTarget default_target;
  // Every target this type will accept. A type that acts on the abiotic state
  // cannot be aimed at a species, and one that acts on individuals cannot be
  // aimed at the environment; stating both directions here is what stops an
  // impossible combination being accepted and then quietly reinterpreted.
  bool allows_patch;
  bool allows_environment;
  bool allows_species;
};

const TypeInfo type_info[] = {
  // amount. The resource is named by target_index, and belongs to the
  // environment rather than to any one species, so a pulse cannot be narrowed
  // to one species.
  {EventType::ResourcePulse,    "resource_pulse",    1,
   EventTarget::Environment, false, true,  false},
  // intensity, duration, threshold, sensitivity
  {EventType::ClimateExtreme,   "climate_extreme",   4,
   EventTarget::Patch,       true,  false, true},
  // fraction, size_min, size_max
  {EventType::Harvest,          "harvest",           3,
   EventTarget::Patch,       true,  false, true},
  {EventType::NodeIntroduction, "node_introduction", 0,
   EventTarget::Species,     false, false, true}
};

const size_t n_type_info = sizeof(type_info) / sizeof(type_info[0]);

const TypeInfo& info_for(EventType type) {
  for (size_t i = 0; i < n_type_info; ++i) {
    if (type == type_info[i].type) {
      return type_info[i];
    }
  }
  util::stop("Unknown event type");
  return type_info[0]; // not reached
}

struct TargetName {
  EventTarget target;
  const char* name;
};

const TargetName target_names[] = {
  {EventTarget::Patch,       "patch"},
  {EventTarget::Environment, "environment"},
  {EventTarget::Species,     "species"}
};

const size_t n_target_names = sizeof(target_names) / sizeof(target_names[0]);

std::string known_type_names() {
  std::string ret;
  for (size_t i = 0; i < n_type_info; ++i) {
    ret += std::string(i > 0 ? ", " : "") + type_info[i].name;
  }
  return ret;
}
}

EventType event_type_from_string(const std::string& name) {
  for (size_t i = 0; i < n_type_info; ++i) {
    if (name == type_info[i].name) {
      return type_info[i].type;
    }
  }
  util::stop("Unknown event type '" + name + "'. Known types: " +
             known_type_names());
  return EventType::NodeIntroduction; // not reached
}

std::string event_type_to_string(EventType type) {
  return info_for(type).name;
}

size_t event_type_n_params(EventType type) {
  return info_for(type).n_params;
}

EventTarget event_type_default_target(EventType type) {
  return info_for(type).default_target;
}

bool event_type_accepts_species(EventType type) {
  return info_for(type).allows_species;
}

bool event_type_allows_target(EventType type, EventTarget target) {
  const TypeInfo& i = info_for(type);
  switch (target) {
  case EventTarget::Patch:       return i.allows_patch;
  case EventTarget::Environment: return i.allows_environment;
  case EventTarget::Species:     return i.allows_species;
  }
  return false;
}

// The targets a type will take, for an error message that says what to do
// rather than only what was wrong.
std::string event_type_target_list(EventType type) {
  const TypeInfo& i = info_for(type);
  std::string ret;
  if (i.allows_patch)       ret += std::string(ret.empty() ? "" : ", ") + "patch";
  if (i.allows_environment) ret += std::string(ret.empty() ? "" : ", ") + "environment";
  if (i.allows_species)     ret += std::string(ret.empty() ? "" : ", ") + "species";
  return ret;
}

EventTarget event_target_from_string(const std::string& name) {
  for (size_t i = 0; i < n_target_names; ++i) {
    if (name == target_names[i].name) {
      return target_names[i].target;
    }
  }
  util::stop("Unknown event target '" + name +
             "'. Known targets: patch, environment, species");
  return EventTarget::Patch; // not reached
}

std::string event_target_to_string(EventTarget target) {
  for (size_t i = 0; i < n_target_names; ++i) {
    if (target == target_names[i].target) {
      return target_names[i].name;
    }
  }
  util::stop("Unknown event target");
  return ""; // not reached
}

namespace {
// Reject a parameter the action cannot act on, here rather than mid-run. Each
// of these is silent otherwise: a NaN intensity becomes a zero-damage climate
// event, a negative sensitivity produces negative applied mortality, and an
// inverted size band harvests nothing at all -- results that look like answers.
void validate_event_params(const std::string& at, EventType type,
                           const std::vector<double>& q) {
  const auto finite = [&](double v, const char* name) {
    if (!util::is_finite(v)) {
      util::stop(at + " has a non-finite " + name);
    }
  };
  const auto nonneg = [&](double v, const char* name) {
    finite(v, name);
    if (v < 0.0) {
      util::stop(at + " has a negative " + name + " (" +
                 util::to_string(v) + ")");
    }
  };
  switch (type) {
  case EventType::ResourcePulse:
    nonneg(q.at(0), "amount");
    break;
  case EventType::ClimateExtreme:
    finite(q.at(0), "intensity");
    nonneg(q.at(1), "duration");
    finite(q.at(2), "threshold");
    nonneg(q.at(3), "sensitivity");
    break;
  case EventType::Harvest: {
    finite(q.at(0), "fraction");
    if (q.at(0) < 0.0 || q.at(0) >= 1.0) {
      util::stop(at + " has fraction " + util::to_string(q.at(0)) +
                 ", which must be in [0, 1): removing all of a cohort would "
                 "take its density to -Inf.");
    }
    nonneg(q.at(1), "size_min");
    // size_max is Inf by default, which is the whole point of the default, so
    // it is bounded rather than required finite.
    if (std::isnan(q.at(2)) || q.at(2) < q.at(1)) {
      util::stop(at + " has size_max " + util::to_string(q.at(2)) +
                 " below size_min " + util::to_string(q.at(1)) +
                 ": the band is empty, so the event would silently do nothing");
    }
    break;
  }
  case EventType::NodeIntroduction:
    break;
  }
}
}

void Events::validate() {
  const size_t n = time.size();
  if (type.size() != n || target.size() != n || target_index.size() != n ||
      params.size() != n) {
    util::stop("Events columns must all have the same length (time has " +
               util::to_string(n) + ")");
  }
  for (size_t i = 0; i < n; ++i) {
    const std::string at = "Event " + util::to_string(i + 1) + " (" +
      type[i] + ")";
    if (!util::is_finite(time[i]) || time[i] < 0.0) {
      util::stop(at + " has a non-finite or negative time");
    }
    // Also rejects an unknown name, with the list of known ones.
    const EventType t = event_type_from_string(type[i]);
    const size_t n_expected = event_type_n_params(t);
    if (params[i].size() != n_expected) {
      util::stop(at + " expects " + util::to_string(n_expected) +
                 " parameters but has " + util::to_string(params[i].size()));
    }
    const EventTarget tg = event_target_from_string(target[i]);
    if (!event_type_allows_target(t, tg)) {
      // Naming the permitted targets matters more than naming the offence: an
      // environment-aimed harvest is a modelling mistake, and the useful reply
      // is what it should have said instead.
      if (tg == EventTarget::Species) {
        util::stop(at + " cannot be aimed at a single species; it accepts " +
                   event_type_target_list(t));
      }
      if (t == EventType::NodeIntroduction) {
        util::stop(at + " must name the species being introduced");
      }
      util::stop(at + " cannot act on the " + target[i] + "; it accepts " +
                 event_type_target_list(t));
    }
    validate_event_params(at, t, params[i]);
  }
}

void validate_event_horizon(const Events& events, double max_time) {
  // An event past the end of the run is not a no-op to be dropped: it usually
  // means a time-unit slip, a truncated run, or a forcing record reused at the
  // wrong length. Silently ignoring it would give a plausible answer that is
  // missing the intervention the user asked for. Caught here, before any
  // integration, rather than surfacing later as the solver's own complaint
  // about being asked to integrate backwards.
  for (size_t i = 0; i < events.size(); ++i) {
    if (events.time[i] > max_time) {
      util::stop("Event " + util::to_string(i + 1) + " (" + events.type[i] +
                 ") occurs at time " + util::to_string(events.time[i]) +
                 ", after max_patch_lifetime = " + util::to_string(max_time) +
                 ". Events must fall within [0, max_patch_lifetime].");
    }
  }
}

std::vector<NodeScheduleEvent> to_schedule_events(const Events& events,
                                                  size_t n_species) {
  std::vector<NodeScheduleEvent> ret;
  ret.reserve(events.size());
  for (size_t i = 0; i < events.size(); ++i) {
    const EventType type = event_type_from_string(events.type[i]);
    const EventTarget target = event_target_from_string(events.target[i]);
    // 1-based on the way in, as everywhere else in the R interface.
    size_t index = 0;
    if (target != EventTarget::Patch) {
      const size_t raw = events.target_index[i];
      if (raw < 1) {
        util::stop("Event " + util::to_string(i + 1) +
                   " has target_index " + util::to_string(raw) +
                   ", which must be at least 1");
      }
      // Species can be bounds-checked here; a resource index cannot, because
      // how many an environment has is the environment's business and it is
      // not in scope. Patch::apply_event() checks it against n_resources().
      if (target == EventTarget::Species && raw > n_species) {
        util::stop("Event " + util::to_string(i + 1) +
                   " has target_index " + util::to_string(raw) +
                   ", outside 1.." + util::to_string(n_species));
      }
      index = raw - 1;
    }
    ret.push_back(NodeScheduleEvent(events.time[i], index, type, target,
                                    events.params[i]));
  }
  return ret;
}

Events events_from_schedule_events(
    const std::vector<NodeScheduleEvent>& events) {
  Events ret;
  ret.time.reserve(events.size());
  ret.type.reserve(events.size());
  ret.target.reserve(events.size());
  ret.target_index.reserve(events.size());
  ret.params.reserve(events.size());
  for (std::vector<NodeScheduleEvent>::const_iterator e = events.begin();
       e != events.end(); ++e) {
    ret.time.push_back(e->time_introduction());
    ret.type.push_back(event_type_to_string(e->type));
    ret.target.push_back(event_target_to_string(e->target));
    ret.target_index.push_back(e->target_index + 1); // back to 1-based
    ret.params.push_back(e->params);
  }
  return ret;
}

EventLog event_log_from_records(const std::vector<EventRecord>& records) {
  EventLog ret;
  ret.time.reserve(records.size());
  ret.type.reserve(records.size());
  ret.target.reserve(records.size());
  ret.target_index.reserve(records.size());
  ret.requested.reserve(records.size());
  ret.applied.reserve(records.size());
  for (std::vector<EventRecord>::const_iterator r = records.begin();
       r != records.end(); ++r) {
    ret.time.push_back(r->time);
    ret.type.push_back(event_type_to_string(r->type));
    ret.target.push_back(event_target_to_string(r->target));
    ret.target_index.push_back(r->target_index + 1);
    ret.requested.push_back(r->requested);
    ret.applied.push_back(r->applied);
  }
  return ret;
}

}
